File: isegoria/src/main/java/com/eulersbridge/isegoria/vote/VoteViewPagerFragment.java
package com.eulersbridge.isegoria.vote;

import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.eulersbridge.isegoria.MainActivity;
import com.eulersbridge.isegoria.R;
import com.eulersbridge.isegoria.util.ui.NonSwipeableViewPager;
import com.eulersbridge.isegoria.util.ui.SimpleFragmentPagerAdapter;
import com.eulersbridge.isegoria.util.ui.TitledFragment;

import java.util.ArrayList;

public class VoteViewPagerFragment extends Fragment
        implements TitledFragment, MainActivity.TabbedFragment {

    private TabLayout tabLayout;

    private SimpleFragmentPagerAdapter viewPagerAdapter;
    private NonSwipeableViewPager viewPager;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.vote_view_pager_fragment, container, false);

        // Ensure options menu from another fragment is not carried over
        if (getActivity() != null)
            getActivity().invalidateOptionsMenu();

        VoteViewModel viewModel = ViewModelProviders.of(this).get(VoteViewModel.class);

        viewModel.locationAndDateComplete.observe(this, complete -> {
            if (complete != null && complete)
                viewPager.setCurrentItem(1);
        });

        viewModel.pledgeComplete.observe(this, complete -> {
            if (complete != null && complete)
                viewPager.setCurrentItem(2);
        });

        setupViewPager(rootView);

        return rootView;
    }

    @Override
    public String getTitle(Context context) {
        return context.getString(R.string.section_title_vote);
    }

    private void setupViewPager(View rootView) {
        if (rootView == null) rootView = getView();

        if (viewPager == null && rootView != null) {
            viewPager = rootView.findViewById(R.id.vote_view_pager_fragment);

            ArrayList<Fragment> fragments = new ArrayList<>();
            fragments.add(new VoteFragment());
            fragments.add(new VotePledgeFragment());
            fragments.add(new VoteDoneFragment());

            viewPagerAdapter = new SimpleFragmentPagerAdapter(getChildFragmentManager(), fragments) {
                @Override
                public CharSequence getPageTitle(int position) {

                    Fragment fragment = viewPagerAdapter.getItem(position);
                    if (fragment instanceof TitledFragment) {
                        return ((TitledFragment) fragment).getTitle(getContext());

                    } else {
                        return null;
                    }
                }
            };

            viewPager.setAdapter(viewPagerAdapter);

            viewPager.setCurrentItem(0);
        }
    }

    @Override
    public void setupTabLayout(TabLayout tabLayout) {
        this.tabLayout = tabLayout;

        tabLayout.removeAllTabs();
        tabLayout.setVisibility(View.VISIBLE);

        tabLayout.setupWithViewPager(viewPager);
        tabLayout.addOnTabSelectedListener(onTabSelectedListener);
        tabLayout.setEnabled(false);
    }

    @Override
    public void onPause() {
        super.onPause();

        if (tabLayout != null) tabLayout.removeOnTabSelectedListener(onTabSelectedListener);
    }

    private final TabLayout.OnTabSelectedListener onTabSelectedListener = new TabLayout.OnTabSelectedListener() {
        @Override
        public void onTabSelected(TabLayout.Tab tab) { }

        @Override
        public void onTabUnselected(TabLayout.Tab tab) { }

        @Override
        public void onTabReselected(TabLayout.Tab tab) { }
    };
}

File: isegoria/src/main/kotlin/com/eulersbridge/isegoria/vote/VoteViewPagerFragment.kt
package com.eulersbridge.isegoria.vote

import android.arch.lifecycle.ViewModelProvider
import android.arch.lifecycle.ViewModelProviders
import android.content.Context
import android.os.Bundle
import android.support.design.widget.TabLayout
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.view.isGone
import com.eulersbridge.isegoria.R
import com.eulersbridge.isegoria.util.extension.observe
import com.eulersbridge.isegoria.util.ui.SimpleFragmentPagerAdapter
import com.eulersbridge.isegoria.util.ui.TabbedFragment
import com.eulersbridge.isegoria.util.ui.TitledFragment
import com.eulersbridge.isegoria.vote.pages.VoteDoneFragment
import com.eulersbridge.isegoria.vote.pages.VoteFragment
import com.eulersbridge.isegoria.vote.pages.VotePledgeFragment
import dagger.android.support.AndroidSupportInjection
import kotlinx.android.synthetic.main.vote_view_pager_fragment.*
import java.util.*
import javax.inject.Inject

class VoteViewPagerFragment : Fragment(), TitledFragment, TabbedFragment {

    @Inject
    internal lateinit var modelFactory: ViewModelProvider.Factory
    private lateinit var viewModel: VoteViewModel

    private lateinit var tabLayout: TabLayout
    private lateinit var viewPagerAdapter: SimpleFragmentPagerAdapter

    override fun onAttach(context: Context?) {
        AndroidSupportInjection.inject(this)
        super.onAttach(context)
        viewModel = ViewModelProviders.of(this, modelFactory)[VoteViewModel::class.java]
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val rootView = inflater.inflate(R.layout.vote_view_pager_fragment, container, false)

        // Ensure options menu from another fragment is not carried over
        activity?.invalidateOptionsMenu()

        observe(viewModel.pageIndex) {
            viewPager.currentItem = it?.value ?: 0
        }

        return rootView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        setupViewPager()
    }

    override fun getTitle(context: Context?) = context?.getString(R.string.section_title_vote)

    private fun setupViewPager() {
        val fragments = ArrayList<Fragment>()

        val voteFragment = VoteFragment()
        voteFragment.setViewModel(viewModel)

        val pledgeFragment = VotePledgeFragment()
        pledgeFragment.setViewModel(viewModel)

        val doneFragment = VoteDoneFragment()
        doneFragment.setViewModel(viewModel)

        fragments.apply {
            add(voteFragment)
            add(pledgeFragment)
            add(doneFragment)
        }

        viewPagerAdapter = object : SimpleFragmentPagerAdapter(childFragmentManager, fragments) {
            override fun getPageTitle(position: Int): CharSequence? {
                val fragment = viewPagerAdapter.getItem(position)
                return if (fragment is TitledFragment) fragment.getTitle(context!!) else null
            }
        }

        viewPager.apply {
            adapter = viewPagerAdapter
            currentItem = 0
        }
    }

    override fun setupTabLayout(tabLayout: TabLayout) {
        this.tabLayout = tabLayout

        tabLayout.apply {
            isGone = true
        }
    }
}